File: Conceitos de LPs/Tipos de Dados/Uniao/uniao.pas
//Pascal
program Uniao;

//A depender do momento o registro Forma vai ter possuir internamente ou um Circulo, ou um Triangulo, ou um Retangulo, selecionado via case, similar ao switch..case.
type
    TForma = record
        case Tipo: (Circulo, Triangulo, Retangulo) of
            Circulo: (Raio: Real);
            Triangulo: (Base, Altura: Real);
            Retangulo: (Largura, Altura: Real);
    end;

var
    Forma: TForma;

begin
    // Usando como Círculo
    Forma.Tipo := Circulo;
    Forma.Raio := 5.0;
    WriteLn('Círculo com raio: ', Forma.Raio:0:2);

    // Usando como Retângulo
    Forma.Tipo := Retangulo;
    Forma.Largura := 10.0;
    Forma.Altura := 20.0;
    WriteLn('Retângulo com largura: ', Forma.Largura:0:2, ' e altura: ', Forma.Altura:0:2);
end.

File: Conceitos de LPs/Tipos de Dados/Uniao/uniao.rs
//Rust
//A depender do momento a União Forma vai ter possuir internamente ou uma estrutura Circulo, ou uma estrutura Triangulo, ou uma estrutura Retangulo.

union Forma {
    circulo: Circulo,
    triangulo: Triangulo,
    retangulo: Retangulo,
}

struct Circulo {
    raio: f32,
}

struct Triangulo {
    base: f32,
    altura: f32,
}

struct Retangulo {
    largura: f32,
    altura: f32,
}

fn main() {
    let forma = Forma {
        circulo: Circulo { raio: 5.0 },
    };

    unsafe {
        // Usando como círculo
        println!("Círculo com raio: {}", forma.circulo.raio);
    }

    let forma = Forma {
        retangulo: Retangulo {
            largura: 10.0,
            altura: 20.0,
        },
    };

    unsafe {
        // Usando como retângulo
        println!("Retângulo com largura: {} e altura: {}", forma.retangulo.largura, forma.retangulo.altura);
    }
}
